static/js/services/generate-search-requests.js

```diff
@@ -139,40 +145,54 @@ var _ = require('underscore'),
         contacts: function(filters) {
-          var typeViews = documentType(filters, 'medic-client/contacts_by_type');
-          var freetextViews = freetext(filters, 'medic-client/contacts_by_freetext');
-          if (typeViews && typeViews.params.keys.length &&
-              freetextViews && freetextViews.length) {
-            return freetextViews.map(function(freetextView) {
+          var typeRequests = documentTypeRequest(filters, 'medic-client/contacts_by_type');
+          // Single request, with one key for each type.
+          var hasTypeRequests = typeRequests && typeRequests.params.keys.length;
+
+          var freetextRequests = freetextRequest(filters, 'medic-client/contacts_by_freetext');
+          // One request per word.
+          var hasFreetextRequests = freetextRequests && freetextRequests.length;
+
+          // If both type and freetext requests,
+          // then we make a special combined contacts+type request.
+          if (hasTypeRequests && hasFreetextRequests) {
+
+            var makeCombinedParams = function(freetextRequest, typeKey) {
+              var type = typeKey[0];
+              var params = {};
+              if (freetextRequest.key) {
+                params.key = [ type, freetextRequest.params.key[0] ];
+              } else {
+                params.startkey = [ type, freetextRequest.params.startkey[0] ];
+                params.endkey = [ type, freetextRequest.params.endkey[0] ];
+              }
+              return params;
+            };
+
+            var makeCombinedRequest = function(freetextRequest, typeRequests) {
               var result = {
                 view: 'medic-client/contacts_by_type_freetext',
-                union: typeViews.params.keys.length > 1
+                union: typeRequests.params.keys.length > 1
               };
+
               if (result.union) {
-                result.params = [];
+                result.paramSets =
+                    typeRequests.params.keys.map(_.partial(makeCombinedParams, freetextRequest, _));
+                return result;
               }
-              typeViews.params.keys.forEach(function(typeKey) {
-                var type = typeKey[0];
-                var params = {};
-                if (freetextView.key) {
-                  params.key = [ type, freetextView.params.key[0] ];
-                } else {
-                  params.startkey = [ type, freetextView.params.startkey[0] ];
-                  params.endkey = [ type, freetextView.params.endkey[0] ];
-                }
-                if (result.union) {
-                  result.params.push(params);
-                } else {
-                  result.params = params;
-                }
-              });
+
+              result.params = makeCombinedParams(freetextRequest, typeRequests.params.keys[0]);
               return result;
-            });
+            };
+
+            return freetextRequests.map(_.partial(makeCombinedRequest, _, typeRequests));
```

Comment:
Done. (though I actually keep the `_`, I like making things explicit, I find it clearer to read)